Add Rainforest RAVEn integration

cottsay opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Add a new integration for the RAVEn family of devices from [Rainforest Automation](https://www.rainforestautomation.com/about-us/), including the [EMU-2](https://www.rainforestautomation.com/rfa-z105-2-emu-2-2/). Though the RAVEn itself isn't sold anymore, the serial protocol for the EMU-2 is based on it and is very nearly identical.

This integration uses the fully-asynchronous [`aioraven`](https://github.com/cottsay/aioraven/) package for communication with the device. The changelog can be found here: https://github.com/cottsay/aioraven/blob/main/CHANGELOG.md

I believe that this new integration meets the criteria for Platinum Quality, but I could be mistaken. Please provide feedback if you disagree.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: home-assistant/home-assistant.io#24505

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [x] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [x] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

cottsay committed:
Add Rainforest RAVEn integration

cottsay committed:
Add Rainforest Automation brand

cottsay committed:
Add diagnostics to Rainforest RAVEn integration

cottsay committed:
Drop a test assertion for an undefined behavior

akx commented on homeassistant/components/rainforest_raven/config_flow.py:
```diff
@@ -0,0 +1,158 @@
+        """Handle USB Discovery."""
+        device = discovery_info.device
+        dev_path = await self.hass.async_add_executor_job(usb.get_serial_by_id, device)
+        unique_id = f"{discovery_info.vid}:{discovery_info.pid}_{discovery_info.serial_number}_{discovery_info.manufacturer}_{discovery_info.description}"
```
This looks like `_generate_unique_id`...

cottsay replied:
This yields the same string but using a different underlying data type. The function operates on `ListPortInfo` but this code utilizes a `UsbServiceInfo` instance. The field names are the same except the UID/GID, which is stored as a number in the former and a string in the latter.

IMO, designing the a function to deal with both data types would result in more code to do the same task.

cottsay replied:
This ended up happening in 9eba9f6.

akx commented on tests/components/rainforest_raven/test_config_flow.py:
```diff
@@ -0,0 +1,334 @@
+
+    result = await hass.config_entries.flow.async_configure(
+        result["flow_id"],
+        user_input={
+            CONF_DEVICE: usb.human_readable_device_name(
+                DISCOVERY_INFO.device,
+                DISCOVERY_INFO.serial_number,
+                DISCOVERY_INFO.manufacturer,
+                DISCOVERY_INFO.description,
+                DISCOVERY_INFO.vid,
+                DISCOVERY_INFO.pid,
+            )
+        },
```
This `user_input` is repeated identically many times in this test file; should it become a constant (like `DISCOVERY_INFO` already is) or a fixture..?

cottsay replied:
Done in 329c19e.

cottsay committed:
Merge branch 'dev' into rainforest_raven

cottsay committed:
Add DEVICE_NAME test constant

cottsay committed:
Merge branch 'dev' into rainforest_raven

edenhaus committed:
Merge branch 'dev' into rainforest_raven

edenhaus commented on homeassistant/components/rainforest_raven/__init__.py:
```diff
@@ -0,0 +1,28 @@
+from . import data
+from .const import DOMAIN
+
+PLATFORMS = ("sensor",)
```
```suggestion
PLATFORMS = [Platform.SENSOR]
```

edenhaus reviewed (changes requested):
Please make sure CI is passing

cottsay committed:
Use Platform.SENSOR

Co-authored-by: Robert Resch <robert@resch.dev>

cottsay committed:
Catch up with reality

cottsay reviewed (commented):
Updated based on changes to `dev` of the past months.

cottsay committed:
Merge branch 'dev' into rainforest_raven

edenhaus commented on homeassistant/components/rainforest_raven/translations/en.json:
```diff
@@ -0,0 +1,27 @@
+{
```
Please remove `translations` folder

cottsay replied:
Done in d4e82a9.

edenhaus commented on homeassistant/components/rainforest_raven/config_flow.py:
```diff
@@ -0,0 +1,159 @@
+from .const import DEFAULT_NAME, DOMAIN
+
+
+def _generate_unique_id(port: ListPortInfo) -> str:
+    """Generate unique id from usb attributes."""
+    return f"{port.vid:04X}:{port.pid:04X}_{port.serial_number}_{port.manufacturer}_{port.description}"
```
```suggestion
def _format_id(value: str | int) -> str:
    if isinstance(value, str):
        return value

    return f"{value:04X}"


def _generate_unique_id(info: ListPortInfo | usb.UsbServiceInfo) -> str:
    """Generate unique id from usb attributes."""
    return "_".join(
        _format_id(info.vid),
        _format_id(info.pid),
        info.serial_number,
        info.manufacturer,
        info.description,
    )
```

cottsay replied:
I did most of this in 9eba9f6, but there are a couple of changes. The VID and PID should be separated by a colon to maintain the pattern used in other extensions, and I needed to handle the `None` value on `usb.UsbServiceInfo` to avoid an exception.

In the end, I kept the f-string over the join.

edenhaus commented on homeassistant/components/rainforest_raven/config_flow.py:
```diff
@@ -0,0 +1,159 @@
+        """Handle USB Discovery."""
+        device = discovery_info.device
+        dev_path = await self.hass.async_add_executor_job(usb.get_serial_by_id, device)
+        unique_id = f"{discovery_info.vid}:{discovery_info.pid}_{discovery_info.serial_number}_{discovery_info.manufacturer}_{discovery_info.description}"
```
```suggestion
        unique_id = _generate_unique_id(discovery_info)
```

Reusing the function reduces the complexity and improves the maintainability and readability.

cottsay replied:
I'm still not convinced this is actually more maintainable, but I took a stab at it in 9eba9f6.

edenhaus commented on homeassistant/components/rainforest_raven/config_flow.py:
```diff
@@ -0,0 +1,159 @@
+
+    async def _validate_device(self, dev_path: str) -> None:
+        self._abort_if_unique_id_configured(updates={CONF_DEVICE: dev_path})
+        async with (
+            async_timeout.timeout(5),
+            RAVEnSerialDevice(dev_path) as raven_device,
+        ):
+            await asyncio.sleep(0.05)
+            for _try in range(3, -1, -1):
+                try:
+                    # Try a few times to communicate with the device,
+                    # allowing any data already in the buffer to flush.
+                    meters = await raven_device.get_meter_list()
+                except ParseError:
+                    if not _try:
+                        raise
+                else:
+                    break
+            for meter in meters.meter_mac_ids:
+                meter_info = await raven_device.get_meter_info(meter=meter)
+                self._meter_macs[meter] = (
+                    "unknown"
+                    if meter_info.meter_type is None
+                    else str(meter_info.meter_type)
+                )
```
This code should go into the third party library

cottsay replied:
I moved the synchronization logic into the library. I'm not sure I agree that the rest of this code should go there. It's opening the device, getting the list of programmed meters, and getting the type of each of those meters. These are all discrete operations and I don't see much value in lumping them together in the third-party library and breaking the one-to-one relationship between library queries and device messages.

edenhaus commented on homeassistant/components/rainforest_raven/config_flow.py:
```diff
@@ -0,0 +1,159 @@
+                await self._validate_device(dev_path)
+            except asyncio.TimeoutError:
+                errors[CONF_DEVICE] = "timeout_connect"
+            except (ParseError, SerialException):
```
Exceptions like these should be handled in the third-party library and the code that raises them.

cottsay replied:
Can you clarify what you're looking to see here? The underlying library needs to express communication errors somehow. Would you like to see errors like this surface under a library-specific exception class?

edenhaus replied:
The library should raise an specific error for these exceptions. These exceptions are implementation details and they should be transparent for the caller.

In our [developers docs](https://developers.home-assistant.io/docs/api_lib_index) we have written the following:

> One of the foundational rules of Home Assistant is that we do not include any protocol specific code. Instead, this code should be put into a standalone Python library and published to PyPI.

cottsay replied:
Okay, the library should only emit RAVEn-specific exceptions now.

edenhaus commented on homeassistant/components/rainforest_raven/data.py:
```diff
@@ -0,0 +1,161 @@
+"""Data update coordination for Rainforest RAVEn devices."""
+from __future__ import annotations
```
Please rename it to `coordinator.py`

cottsay replied:
Done in 99ac2cf.

edenhaus commented on homeassistant/components/rainforest_raven/sensor.py:
```diff
@@ -0,0 +1,204 @@
+    async_add_entities(entities)
+
+
+class RAVEnSensor(CoordinatorEntity, SensorEntity):
```
```suggestion
class RAVEnSensor(CoordinatorEntity[RAVEnDataCoordinator], SensorEntity):
```

For new integrations, we require the following:
- Support for [the new entity naming](https://developers.home-assistant.io/docs/core/entity#has_entity_name-true-mandatory-for-new-integrations)
- Services must be [translatable](https://developers.home-assistant.io/blog/2023/07/11/translating-services)
- Entities and their attributes should be [translatable](https://developers.home-assistant.io/blog/2023/03/27/entity_name_translations/)

Please adopt your code to add support for it.

cottsay replied:
Added translations in d4e82a9. Still working on the typing change here to `CoordinatorEntity[RAVEnDataCoordinator]`.

cottsay replied:
Alright, the typing change is done now too.

cottsay committed:
Make rainforest_raven translatable

cottsay committed:
Stop setting device_class on unsupported scenarios

cottsay committed:
Rename rainforest_raven.data -> rainforest_raven.coordinator

cottsay committed:
Make _generate_unique_id more reusable

cottsay committed:
Move device synchronization into third party library

cottsay committed:
Merge branch 'dev' into rainforest_raven

edenhaus commented on homeassistant/components/rainforest_raven/config_flow.py:
```diff
@@ -0,0 +1,156 @@
+            {
+                vol.Required(
+                    CONF_MAC, description={"suggested_value": ",".join(discovered_macs)}
+                ): str
```
Please use a SelectSelector instead

cottsay replied:
Done in bdce69a.

edenhaus commented on homeassistant/components/rainforest_raven/manifest.json:
```diff
@@ -0,0 +1,26 @@
+  "dependencies": ["usb"],
+  "documentation": "https://www.home-assistant.io/integrations/rainforest_raven",
+  "iot_class": "local_polling",
+  "requirements": ["aioraven==0.4.0", "iso4217==1.11.20220401"],
```
```suggestion
  "requirements": ["aioraven==0.4.0"],
```

Your integration code is not using `iso4217`

edenhaus commented on homeassistant/components/rainforest_raven/config_flow.py:
```diff
@@ -0,0 +1,156 @@
+    def __init__(self) -> None:
+        """Set up flow instance."""
+        self._dev_path: str | None = None
+        self._meter_macs: dict[bytes, str] = {}
```
```suggestion
        self._meter_macs: set[str] = set()
```

edenhaus commented on homeassistant/components/rainforest_raven/config_flow.py:
```diff
@@ -0,0 +1,156 @@
+            meters = await raven_device.get_meter_list()
+            for meter in meters.meter_mac_ids:
+                meter_info = await raven_device.get_meter_info(meter=meter)
+                self._meter_macs[meter] = (
+                    "unknown"
+                    if meter_info.meter_type is None
+                    else str(meter_info.meter_type)
+                )
```
```suggestion
                if meter_info.meter_type is None or meter_info.meter_type == MeterType.ELECTRIC:
                    self._meter_macs.add(meter.hex())
```

edenhaus commented on homeassistant/components/rainforest_raven/config_flow.py:
```diff
@@ -0,0 +1,156 @@
+        errors: dict[str, str] = {}
+        if user_input is not None:
+            meter_macs = []
+            for raw_mac in user_input.get(CONF_MAC, "").split(","):
+                try:
+                    mac = bytes.fromhex(raw_mac)
+                except ValueError:
+                    errors[CONF_MAC] = "invalid_mac"
+                    break
+                if len(mac) != 8:
+                    errors[CONF_MAC] = "invalid_mac"
+                    break
+                if self._meter_macs.get(mac) not in ("unknown", "electric"):
+                    errors[CONF_MAC] = "no_meters_found"
+                    break
+                if mac not in meter_macs:
+                    meter_macs.append(mac)
+            if meter_macs and not errors:
+                return self.async_create_entry(
+                    title=user_input.get(CONF_NAME, DEFAULT_NAME),
+                    data={
+                        CONF_DEVICE: self._dev_path,
+                        CONF_MAC: [m.hex() for m in meter_macs],
+                    },
+                )
```
Can be simplified if using a SelectSelector

cottsay replied:
I believe that `SelectSelector` means that only a single value can be chosen, right? The RAVEn can be connected to multiple meters (though I've never tested it).

If you're okay with that limitation, I can switch to the `SelectSelector` anyway.

emontnemery replied:
You need to set the `multiple` flag to `True` on the `SelectSelector`

cottsay replied:
Thanks, @emontnemery.

This was implemented in bdce69a.

edenhaus commented on homeassistant/components/rainforest_raven/config_flow.py:
```diff
@@ -0,0 +1,156 @@
+                    },
+                )
+
+        discovered_macs = [m.hex() for m in self._meter_macs]
```
```suggestion
        discovered_macs = [m for m in self._meter_macs]
```

cottsay replied:
Done as part of adf3027.

cottsay committed:
Merge branch 'dev' into rainforest_raven

cottsay committed:
Switch from asyncio_timeout to asyncio.timeout

cottsay committed:
Ignore non-electric meters

Co-authored-by: Robert Resch <robert@resch.dev>

cottsay committed:
Drop direct dependency on iso4217, bump aioraven

cottsay committed:
Use RAVEn-specific exceptions

cottsay committed:
Add timeouts to data updates

cottsay committed:
Move DeviceInfo generation from Sensor to Coordinator

cottsay committed:
Store meter macs as strings

cottsay commented:
Okay, I've addressed most of @edenhaus' points and I think this is ready for another round of review.

cottsay committed:
Merge branch 'dev' into rainforest_raven

emontnemery commented:
I'm setting this back to draft. Please mark it as ready for review when you've updated to use `SelectSelector` 👍

cottsay committed:
Convert to using SelectSelector

cottsay committed:
Merge branch 'dev' into rainforest_raven

cottsay commented:
Ready for review once again.

---

I've been using a fake device to test the integration: https://github.com/cottsay/aioraven/blob/main/test/mock_device.py

You can set up the device using socat: `sudo socat -d -d pty,raw,echo=0,link=/dev/ttyACM0 "system:python3 test/mock_device.py,pipes"`

It doesn't set the appropriate PID/VID for USB discovery to work, but you can manually add the integration. It was pretty useful for ensuring that the error handling in the integration works as expected.

emontnemery commented on tests/components/rainforest_raven/test_config_flow.py:
```diff
@@ -282,9 +254,7 @@ async def test_flow_user_invalid_type(
-    assert result
-    assert result.get("type") == data_entry_flow.FlowResultType.FORM
-    assert result.get("errors") == {CONF_MAC: "no_meters_found"}
+    with pytest.raises(voluptuous.error.MultipleInvalid):
+        result = await hass.config_entries.flow.async_configure(
+            result["flow_id"], user_input={CONF_MAC: ["abcdef1234567890"]}
+        )
```
When can this happen?

cottsay replied:
I'd suppose it no longer can. I was just adapting a previous test case to work correctly under the `SelectSelector` change.

I'll go ahead and drop it now.

cottsay replied:
Done in 19bc827.

cottsay committed:
Drop test_flow_user_invalid_mac

This test isn't necessary now that SelectSelector is used.

cottsay committed:
Merge branch 'dev' into rainforest_raven

emontnemery commented on homeassistant/components/rainforest_raven/config_flow.py:
```diff
@@ -0,0 +1,155 @@
+
+def _generate_unique_id(info: ListPortInfo | usb.UsbServiceInfo) -> str:
+    """Generate unique id from usb attributes."""
+    return f"{_format_id(info.vid)}:{_format_id(info.pid)}_{info.serial_number}_{info.manufacturer}_{info.description}"
```
Please split this, for example:
```suggestion
    return (
        f"{_format_id(info.vid)}:{_format_id(info.pid)}_{info.serial_number}"
        f"_{info.manufacturer}_{info.description}"
    )
```

cottsay replied:
Implemented in c142a16.

emontnemery commented on homeassistant/components/rainforest_raven/diagnostics.py:
```diff
@@ -0,0 +1,45 @@
+    if not data.get("Meters"):
+        return data
+
+    redacted = {**data, "Meters": {**data["Meters"]}}
```
This is not a very idiomatic to selectively copy the `Meters` dictionary, please consider using `copy.deepcopy` or add a comment.

The code works now, but it could break later if we change what's being mutated.

cottsay replied:
That's fair, I added a comment in c142a16.

emontnemery replied:
I think this suggestion by @ludeeus is much clearer:

```py
    redacted = {**data, "Meters": {}}
    for idx, mac_id in enumerate(data["Meters"]):
        redacted["Meters"][f"**REDACTED{idx}**"] = data["Meters"][mac_id]
```

cottsay replied:
Agreed, implemented in b28067d.

emontnemery commented on homeassistant/components/rainforest_raven/sensor.py:
```diff
@@ -0,0 +1,193 @@
+        key="summation_delivered",
+        native_unit_of_measurement=UnitOfEnergy.KILO_WATT_HOUR,
+        device_class=SensorDeviceClass.ENERGY,
+        state_class=SensorStateClass.TOTAL_INCREASING,
```
Is `TOTAL_INCREASING` the right state class? Why not `TOTAL`?

cottsay replied:
The same state class is used in the Rainforest Eagle integration: https://github.com/home-assistant/core/blob/63136572a54c9b443a1eea25bc22334c1009b143/homeassistant/components/rainforest_eagle/sensor.py#L34

My impression was that `TOTAL` would be good for a value with a known or static time window, while `TOTAL_INCREASING` would be better for a value which just keeps going up but may be reset to zero at some arbitrary point. I would think that a utility meter better fits the latter.

emontnemery replied:
Sure, if the reading resets to 0 at arbitrary points in time, `TOTAL_INCREASING` should be used 👍

emontnemery commented on homeassistant/components/rainforest_raven/sensor.py:
```diff
@@ -0,0 +1,193 @@
+        key="summation_received",
+        native_unit_of_measurement=UnitOfEnergy.KILO_WATT_HOUR,
+        device_class=SensorDeviceClass.ENERGY,
+        state_class=SensorStateClass.TOTAL_INCREASING,
```
Same question.

emontnemery commented on homeassistant/components/rainforest_raven/sensor.py:
```diff
@@ -0,0 +1,193 @@
+from .coordinator import RAVEnDataCoordinator
+
+
+@dataclass
```
```suggestion
@dataclass(frozen=True)
```

cottsay replied:
That doesn't seem to work:
```
error: Cannot inherit frozen dataclass from a non-frozen one
```

emontnemery replied:
It should work now after you merged in dev to your PR branch, please give it another try.

cottsay replied:
Ah, I understand. Thanks, it's working now. Implemented in 7317d54.

emontnemery commented on homeassistant/components/rainforest_raven/sensor.py:
```diff
@@ -0,0 +1,193 @@
+    @property
+    def unique_id(self) -> str:
+        """Return unique ID of entity."""
+        return f"{self._source_mac_address}.{self.entity_description.message_key}.{self.entity_description.key}"
```
Why is this not set in the constructor:
```
self._attr_unique_id = f"{self._source_mac_address}.{self.entity_description.message_key}.{self.entity_description.key}"
```

Also, please line break

cottsay committed:
Implement PR feedback

- Split some long format lines
- Simplify meter mac_id extraction in diagnostics
- Expose unique_id using an attribute instead of a property
- Add a comment about the meters dictionary shallow copy

Co-authored-by: Erik Montnemery <erik@montnemery.com>

cottsay committed:
Merge branch 'dev' into rainforest_raven

cottsay commented on homeassistant/components/rainforest_raven/diagnostics.py:
```diff
@@ -0,0 +1,45 @@
+
+    for idx, mac_id in enumerate(data["Meters"]):
+        redacted["Meters"][f"**REDACTED{idx}**"] = redacted["Meters"][mac_id]
+        del redacted["Meters"][mac_id]
```
Implemented in c142a16.

cottsay committed:
Simplify mac address redaction

Co-authored-by: Joakim Sørensen <ludeeus@ludeeus.dev>

cottsay committed:
Merge branch 'dev' into rainforest_raven

cottsay committed:
Merge branch 'dev' into rainforest_raven

cottsay committed:
Freeze RAVEnSensorEntityDescription dataclass

Co-authored-by: Erik Montnemery <erik@montnemery.com>

emontnemery reviewed (approved):
LGTM, thanks @cottsay 👍

MartinHjelmare commented on tests/components/rainforest_raven/test_diagnostics.py:
```diff
@@ -0,0 +1,103 @@
+    """Mock a functioning RAVEn device."""
+    mock_device = create_mock_device()
+    with patch(
+        "homeassistant.components.rainforest_raven.coordinator.RAVEnSerialDevice",
```
This looks like a copy from the coordinator tests. Common fixtures should go in a `conftest.py` module.

MartinHjelmare reviewed (commented):
Please address the comments in a new PR. Thanks!

MartinHjelmare commented on homeassistant/components/rainforest_raven/coordinator.py:
```diff
@@ -0,0 +1,163 @@
+        )
+
+    @property
+    def device_fw_version(self) -> str | None:
```
If these properties are only used to set the device info, we don't need them. We already check `self._device_info` in the `device_info` property.

MartinHjelmare commented on homeassistant/components/rainforest_raven/coordinator.py:
```diff
@@ -0,0 +1,163 @@
+    @property
+    def device_name(self) -> str:
+        """Return the product name of the device."""
+        return "RAVEn Device"
```
We don't need a property for a constant.

MartinHjelmare commented on homeassistant/components/rainforest_raven/coordinator.py:
```diff
@@ -0,0 +1,163 @@
+        return None
+
+    @property
+    def device_mac_address(self) -> str | None:
```
This property is at least used in two places. If the mac address isn't expected to change we can just use a regular attribute to store the mac address though, when setting `self_device_info`.
